go/pvl/parse.go

```diff
@@ -0,0 +1,240 @@
+// Copyright 2016 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package pvl
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/keybase/client/go/protocol/keybase1"
+)
+
+func parse(in string) (pvlT, error) {
+	b := []byte(in)
+	p := pvlT{}
+	p.PvlVersion = -1
+	p.Revision = -1
+	err := json.Unmarshal(b, &p)
+	if p.PvlVersion == -1 {
+		return p, fmt.Errorf("pvl_version required")
+	}
+	if p.Revision == -1 {
+		return p, fmt.Errorf("revision required")
+	}
+	return p, err
+}
+
+type pvlT struct {
+	PvlVersion int `json:"pvl_version"`
+	Revision   int `json:"revision"`
+	// services is a map from service to a list of scripts.
+	// each script is a list of instructions.
+	Services servicesT `json:"services"`
+}
+
+type servicesT struct {
+	Map map[keybase1.ProofType][]scriptT
+}
+
+func (x *servicesT) UnmarshalJSON(b []byte) error {
+	// read as string map
+	m := make(map[string][]scriptT)
+	err := json.Unmarshal(b, &m)
+	if err != nil {
+		return err
+	}
+	// copy to ProofType map
+	x.Map = make(map[keybase1.ProofType][]scriptT)
+	for k, v := range m {
+		t, ok := keybase1.ProofTypeMap[strings.ToUpper(k)]
+		if ok {
+			x.Map[t] = v
+		}
+		// Unrecognized proof types are discarded silently
+		// So that old clients don't break if a new service is added
+	}
+	return nil
+}
+
+type scriptT struct {
+	Instructions []instructionT
+}
+
+func (x *scriptT) UnmarshalJSON(b []byte) error {
+	err := json.Unmarshal(b, &x.Instructions)
+	for _, ins := range x.Instructions {
+		if ins.variantsFilled() != 1 {
+			return fmt.Errorf("exactly 1 variant must appear in instruction")
+		}
+	}
+	return err
+}
+
+type instructionT struct {
+	// Exactly one of these shall be non-nil
+	// This list is duplicated to:
+	// - instructionT.variantsFilled
+	// - instructionT.String
+	// - stepInstruction
+	// This invariant is enforced by scriptT.UnmarshalJSON
+	AssertRegexMatch    *assertRegexMatchT    `json:"assert_regex_match,omitempty"`
+	AssertFindBase64    *assertFindBase64T    `json:"assert_find_base64,omitempty"`
+	WhitespaceNormalize *whitespaceNormalizeT `json:"whitespace_normalize,omitempty"`
+	RegexCapture        *regexCaptureT        `json:"regex_capture,omitempty"`
+	Fetch               *fetchT               `json:"fetch,omitempty"`
+	SelectorJSON        *selectorJSONT        `json:"selector_json,omitempty"`
+	SelectorCSS         *selectorCSST         `json:"selector_css,omitempty"`
+	TransformURL        *transformURLT        `json:"transform_url,omitempty"`
+}
+
+func (ins *instructionT) variantsFilled() int {
+	n := 0
+	if ins.AssertRegexMatch != nil {
+		n += 1
+	}
+	if ins.AssertFindBase64 != nil {
+		n += 1
+	}
+	if ins.WhitespaceNormalize != nil {
+		n += 1
+	}
+	if ins.RegexCapture != nil {
+		n += 1
+	}
+	if ins.Fetch != nil {
+		n += 1
+	}
+	if ins.SelectorJSON != nil {
+		n += 1
+	}
+	if ins.SelectorCSS != nil {
+		n += 1
+	}
+	if ins.TransformURL != nil {
+		n += 1
+	}
```

Comment:
This is the sort of thing that's easy to forget about if we add new fields above. Any chance we could replace it by just checking that there's only one key in the dict? I guess that would require a second unmarshal, which is annoying, but it might be better than hoping we maintain this function.